### PR TITLE
Auto-complete first and last name when using Link PM

### LINKED ISSUE #2444: Stripe Link: Pre-fill first and last names on checkout
**Description**
Similar to how Stripe Link pre-fills user addresses, we should implement the feature to pre-fill the first and last names on the checkout page.

**Acceptance criteria**
First and last names should be pre-filled on checkout along with other user details.

### PR DESCRIPTION
**Pull request by dpaun1985**
Fixes #2444

## Changes proposed in this Pull Request:
This PR updates the Link functionality, in order to pre-fill first and last name fields on the checkout form, along with other user details.

## Testing instructions
* Navigate to Settings -> Payment Methods and enable Link payment method.
* create an order, go to checkout page and complete all checkout fields. Select also 1click payment under credit card fields
* finish the order
* create another order, go to checkout page and fill the email
*  a pop-up should be shown to type the pin ( 0 0 0 0 0 0 for tests )
* first and last name should be filled for both billing and shipping address
* check if first and last name fields are filled also for blocks checkout page
* test the above scenarios for all options in `Woo -> Settings -> Shipping -> Shipping options -> Shipping destination`

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

**Comment by kalessil** [comment]
Looks like we are porting the changes from WCPay, looks good.
One question, though: will the autofill of first/last name be applied for both the checkout page and blocks?

File: client/stripe-link/index.js [hunk 1]
```diff
@@ -52,39 +52,75 @@ const enableStripeLinkPaymentMethod = ( options ) => {
...
+				shippingAddress,
+				options.shipping_fields.postal_code,
+				'postal_code'
 			);
 		}
 
 		if ( options.complete_billing() ) {
+			const billingNames = billingAddress.name.split( / (.*)/s, 2 );
+			billingAddress.address.last_name = billingNames[ 1 ];
+			billingAddress.address.first_name = billingNames[ 0 ];
```
Comment (reykjalin) [inline client/stripe-link/index.js]: Is there a chance `billingAddress.name` could be just one name? There are certainly people with just 1 name out there as per [falsehoods people believe about names](https://shinesolutions.com/2018/01/08/falsehoods-programmers-believe-about-names-with-examples/). There are also alphabets that don't use spaces in names, such as (as far as I understand it) both Chinese and Japanese names.

If someone like that is paying then `billingNames[ 1 ]` will return `undefined`, so I think we should make sure we handle this case with a `?? ''`:

```suggestion
			const billingNames = billingAddress.name.split( / (.*)/s, 2 );
			billingAddress.address.last_name = billingNames[ 1 ] ?? '';
			billingAddress.address.first_name = billingNames[ 0 ];
```

On a separate (and probably less serious) note people can have multiple first names just as they can have multiple last names. Is there any way we can infer that from the information available here? If not, I think taking the first "word" in a name (like you've already done) is a decent assumption to make — although sometimes titles are included in first names which could make this confusing. However, if we do have access to better information it would probably be best to use that.

**To recap:** The implementation is great as is, provided we don't have access to any further information on the name, such as which part the customer typed in as their given names and surnames 🙂

**Comment by dpaun1985** [comment]
> Looks like we are porting the changes from WCPay, looks good. One question, though: will the autofill of first/last name be applied for both the checkout page and blocks?

Yes, it should work also for blocks checkout page.

**Review by reykjalin — CHANGES_REQUESTED** [review]
Looks good and tests well, but I had some questions regarding the logic used to split a name into a "first name" and "last name" given that names can be difficult to deal with correctly.

Other than that this is looking good 🙂

File: client/stripe-link/index.js (hunk 1 above)
Comment (dpaun1985) [reply client/stripe-link/index.js]: > If someone like that is paying then billingNames[ 1 ] will return undefined, so I think we should make sure we handle this case with a ?? '':
fixed in [eb6f92f](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2457/commits/eb6f92fc9cd80639dfb496080c1912408228271f) . thank you!

>On a separate (and probably less serious) note people can have multiple first names just as they can have multiple last names. Is there any way we can infer that from the information available here? If not, I think taking the first "word" in a name (like you've already done) is a decent assumption to make — although sometimes titles are included in first names which could make this confusing. However, if we do have access to better information it would probably be best to use that.

At this point Stripe does not share any other information.

Thank you for the review .

**Review by reykjalin — APPROVED** [review]
Looks good and tests well! 🚀

File: client/stripe-link/index.js [hunk 2]
```diff
@@ -52,39 +52,75 @@ const enableStripeLinkPaymentMethod = ( options ) => {
...
+			);
+			jQuery(
+				'#billing_country, #billing_state, #shipping_country, #shipping_state'
+			).trigger( 'change' );
+			fillWith( shippingAddress, options.shipping_fields.state, 'state' );
+			fillWith(
+				shippingAddress,
+				options.shipping_fields.postal_code,
+				'postal_code'
 			);
```
Comment (KarlisJ) [inline client/stripe-link/index.js:85]: Without knowledge of codebase I don't understand 
- what this block does
- why 
- how is that related to the changes of this PR

Same comment about block below.

**Review by KarlisJ — COMMENTED** [review]
LGTM! Left one comment that is purely nitpick about code quality/readability. 

If there's good reason to ignore it, I'm happy to approve PR.

File: client/stripe-link/index.js (hunk 2 above)
Comment (dpaun1985) [reply client/stripe-link/index.js:85]: this block addresses the state and postal code field. I did a small refactor, and replaced the `jQuery( 'select' ).trigger( 'change' );` that has a generic selector with a more targeted one.

**Review by kalessil — APPROVED** [review]
LGTM!
